### PR TITLE
Create separate GSI monitor jobs and remove from vrfy job

### LINKED ISSUE #1908: Create separate GSI monitor jobs
### What new functionality do you need?

The GSI monitor jobs (radmon, ozmon, minmon) currently run within the vrfy job. They need to be their own separate jobs. Break them out of the vrfy job as part of work for issue #235.

### What are the requirements for the new functionality?

New, separate, toggle-able jobs in the rocoto mesh for each GSI monitor task.

### Acceptance Criteria

Three new functional jobs.

### TO DO

- [x] Remove GSI monitor tasks from vrfy job
- [x] Remove GSI monitor settings from `config.vrfy`
- [x] Create rocoto job scripts for GSI monitor jobs
- [x] Add GSI monitor jobs to setup scripts
- [x] Update output location to `COM` and update JJOB scripts to use `COM_*` variables
- [x] Add GSI monitor job output to archival

### PR DESCRIPTION
**Pull request by KateFriedman-NOAA**
# Description

This PR will move the GSI monitor tasks from the `vrfy` job into their own separate jobs in the rocoto mesh. This takes care of part of issue #235.

Changes in this PR:

1. Create new individual GSI monitor jobs in the rocoto mesh to match how it is done in operations. The new jobs are: `gdasverfozn`, `gdasverfrad`, `gdasvminmon`, and `gfsvminmon`. The names correspond to the names in operations and the ecf script names. The jobs are only added to cycled mode, none of the jobs are run in forecast-only mode.
2. Updates to GSI monitor scripts:
  * Update `ush/jjob_header.sh` arguments in JJOB scripts to call new individual job names and configs.
  * Introduce `COM_*` variables and use of the `generate_com` utility. This results in the outputs landing in `COM` subfolders, like occurs in operations, instead of landing in the online archive.
  * Some variable cleanup in JJOB and ush scripts, includes cleaning up usage of `DATA` variable and its removal at the end of the jobs.
  * Add missing exit statements to `JGFS_ATMOS_VMINMON` and `JGDAS_ATMOS_VERFRAD`.
3. Create new rocoto job scripts for the new individual GSI monitor jobs.
4. Create new configs for the new individual GSI monitor jobs.
5. Add new individual GSI monitor jobs to resource configuration, using the same resources as in operations.
6. Move and rename GSI monitor job switches from `config.vrfy` to `config.base`. Also add switches to setup system. New switches to control each job being added to the rocoto mesh are: `DO_VERFOZN`, `DO_VERFRAD`, `DO_VMINMON`
7. Add GSI monitor job output to archival job and into resulting HPSS tarballs or local tarballs. This now matches what is done in operations.
8. Remove the GSI monitor tasks from the rocoto `vrfy` job and associated `config.vrfy`.

Example output locations and files (not showing contents of oznmon/horiz or oznmon/time for brevity, see archival script for full list of output files):
```
kate.friedman@dlogin09:/lfs/h2/emc/ptmp/kate.friedman/comrot/testmonitor2> ll gdas.20211221/00/products/atmos/*mon
gdas.20211221/00/products/atmos/minmon:
total 112
-rw-r--r-- 1 kate.friedman emc 14030 Oct 25 15:00 2021122100.costs.txt
-rw-r--r-- 1 kate.friedman emc 84932 Oct 25 15:00 2021122100.cost_terms.txt
-rw-r--r-- 1 kate.friedman emc   808 Oct 25 15:00 2021122100.gnorms.ieee_d
-rw-r--r-- 1 kate.friedman emc   808 Oct 25 15:00 2021122100.reduction.ieee_d
-rw-r--r-- 1 kate.friedman emc    80 Oct 25 15:00 gnorm_data.txt

gdas.20211221/00/products/atmos/oznmon:
total 8
drwxr-sr-x 2 kate.friedman emc 4096 Oct 25 15:05 horiz
drwxr-sr-x 2 kate.friedman emc 4096 Oct 25 15:05 time

gdas.20211221/00/products/atmos/radmon:
total 21036
-rw-r--r-- 1 kate.friedman emc      231 Oct 25 15:12 bad_diag.2021122100
-rw-r--r-- 1 kate.friedman emc     9035 Oct 25 15:12 bad_pen.2021122100
-rw-r--r-- 1 kate.friedman emc     1449 Oct 25 15:12 low_count.2021122100
-rw-r--r-- 1 kate.friedman emc 20523403 Oct 25 15:07 radmon_angle.tar.gz
-rw-r--r-- 1 kate.friedman emc   217272 Oct 25 15:08 radmon_bcoef.tar.gz
-rw-r--r-- 1 kate.friedman emc   502151 Oct 25 15:10 radmon_bcor.tar.gz
-rw-r--r-- 1 kate.friedman emc   264480 Oct 25 15:12 radmon_time.tar.gz
-rw-r--r-- 1 kate.friedman emc      684 Oct 25 15:12 warning.2021122100
kate.friedman@dlogin09:/lfs/h2/emc/ptmp/kate.friedman/comrot/testmonitor2> ll gfs.20211221/00/products/atmos/minmon/
total 88
-rw-r--r-- 1 kate.friedman emc 10530 Oct 25 14:55 2021122100.costs.txt
-rw-r--r-- 1 kate.friedman emc 63882 Oct 25 14:55 2021122100.cost_terms.txt
-rw-r--r-- 1 kate.friedman emc   808 Oct 25 14:55 2021122100.gnorms.ieee_d
-rw-r--r-- 1 kate.friedman emc   608 Oct 25 14:55 2021122100.reduction.ieee_d
-rw-r--r-- 1 kate.friedman emc    80 Oct 25 14:55 gnorm_data.txt
```

Lingering work and considerations:

1. Issue #1925
2. Consider not hardcoding the list of types in archival and perhaps tie it to something reliable or configurable (maybe into `config.verfozn`?):
`subtyplist="gome_metop-b omi_aura ompslp_npp ompsnp_n20 ompsnp_npp ompstc8_n20 ompstc8_npp sbuv2_n19"`

Resolves #1908

# Type of change

- New feature (adds functionality)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? NO

# How has this been tested?

- Two cycled 2.5 cycle atmos-only tests on WCOSS2
- Set up forecast-only test to confirm jobs not added to rocoto mesh, did not run forecast-only test

# Checklist
- [x] Any dependent changes have been merged and published
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [x] I have made corresponding changes to the documentation if necessary

### CODE REVIEW

**Comment by KateFriedman-NOAA** [comment]
Shellcheck warnings resolved and a couple missing `exit` statements added to JJOB scripts (found during resolving shellcheck warnings). Ready for review!

**Review by EdwardSafford-NOAA — APPROVED** [review]
Looks good to me.

File: jobs/JGDAS_ATMOS_VERFRAD
Comment (WalterKolczynski-NOAA) [inline jobs/JGDAS_ATMOS_VERFRAD]: The variable name says previous, but the YMD and HH used are the current cycle

Comment (WalterKolczynski-NOAA) [reply jobs/JGDAS_ATMOS_VERFRAD]:
```suggestion
YMD=${P_PDY} HH=${p_cyc} generate_com -rx COM_ATMOS_RADMON_PREV:COM_ATMOS_RADMON_TMPL
```

File: parm/config/gfs/config.resources
Comment (WalterKolczynski-NOAA) [inline parm/config/gfs/config.resources:748]: The job is probably going away anyway, but I'm fairly certain it doesn't require 5 hr anymore to make a few wgrib2 calls.

File: jobs/JGDAS_ATMOS_VERFRAD
Comment (WalterKolczynski-NOAA) [inline jobs/JGDAS_ATMOS_VERFRAD:55]: Doesn't have to be this PR, but I would love to retire these intermediate variables and use the COM variables directly. (I already did this for most variables during the initial COM refactor.)

File: workflow/applications/gfs_cycled.py
Comment (WalterKolczynski-NOAA) [inline workflow/applications/gfs_cycled.py:180]: Duplicated line
```suggestion
```

File: workflow/rocoto/gfs_tasks.py
Comment (WalterKolczynski-NOAA) [inline workflow/rocoto/gfs_tasks.py:1]: The dependency of the archive job needs to be updated to wait for the new GSI verification jobs.

**Review by WalterKolczynski-NOAA — CHANGES_REQUESTED** [review]

File: workflow/applications/gfs_cycled.py
Comment (KateFriedman-NOAA) [reply workflow/applications/gfs_cycled.py:180]: Thanks for catching!

File: jobs/JGDAS_ATMOS_VERFRAD
Comment (KateFriedman-NOAA) [reply jobs/JGDAS_ATMOS_VERFRAD]: Thanks for catching!

File: parm/config/gfs/config.resources
Comment (KateFriedman-NOAA) [reply parm/config/gfs/config.resources:748]: As of this PR it still runs the tracker, genesis, and MOS tasks as well. I will reduce it in the next PR that removes the tracker and genesis out of vrfy. I have added a checklist item to issue #1988.

File: jobs/JGDAS_ATMOS_VERFRAD
Comment (KateFriedman-NOAA) [reply jobs/JGDAS_ATMOS_VERFRAD:55]: I could tackle this in follow-up issue #1925.

File: workflow/rocoto/gfs_tasks.py
Comment (KateFriedman-NOAA) [reply workflow/rocoto/gfs_tasks.py:1]: The `verfozn` and `verfrad` jobs both need the `analdiag` job complete before they run and the `gdas[gfs]vminmon` job needs `gdas[gfs]anal` job complete. The `verfozn` job ran in 23s (5m walltime), the `verfrad` in 441s (20m walltime), and `vminmon` in 18s (5m walltime) in my testing. So in theory they should all be done long before archive runs.

Here are the log timestamps for the gdas jobs (I reran the monitor jobs to their logs have .0 now):
```
kate.friedman@dlogin09:/lfs/h2/emc/ptmp/kate.friedman/comrot/testmonitor2> ls -ltr logs/2021122106/gdas*
-rw-r--r-- 1 kate.friedman emc 13567475 Oct 25 19:10 logs/2021122106/gdasprep.log
-rw-r--r-- 1 kate.friedman emc   836509 Oct 25 19:34 logs/2021122106/gdasanal.log
-rw-r--r-- 1 kate.friedman emc    67357 Oct 25 19:35 logs/2021122106/gdasvminmon.log.0
-rw-r--r-- 1 kate.friedman emc   406080 Oct 25 19:35 logs/2021122106/gdassfcanl.log
-rw-r--r-- 1 kate.friedman emc  1874474 Oct 25 19:36 logs/2021122106/gdasanaldiag.log
-rw-r--r-- 1 kate.friedman emc   121140 Oct 25 19:40 logs/2021122106/gdasverfozn.log.0
-rw-r--r-- 1 kate.friedman emc   183405 Oct 25 19:40 logs/2021122106/gdasanalcalc.log
-rw-r--r-- 1 kate.friedman emc  1297324 Oct 25 19:46 logs/2021122106/gdaspost_f000.log
-rw-r--r-- 1 kate.friedman emc  1400286 Oct 25 19:46 logs/2021122106/gdaspost_f003.log
-rw-r--r-- 1 kate.friedman emc  1400395 Oct 25 19:46 logs/2021122106/gdaspost_f001.log
-rw-r--r-- 1 kate.friedman emc  1400405 Oct 25 19:46 logs/2021122106/gdaspost_f002.log
-rw-r--r-- 1 kate.friedman emc   471080 Oct 25 19:47 logs/2021122106/gdasverfrad.log.0
-rw-r--r-- 1 kate.friedman emc  1257264 Oct 25 19:48 logs/2021122106/gdasfcst.log
-rw-r--r-- 1 kate.friedman emc  1252659 Oct 25 19:51 logs/2021122106/gdaspost_anl.log
-rw-r--r-- 1 kate.friedman emc  1400479 Oct 25 19:51 logs/2021122106/gdaspost_f009.log
-rw-r--r-- 1 kate.friedman emc  1399989 Oct 25 19:51 logs/2021122106/gdaspost_f007.log
-rw-r--r-- 1 kate.friedman emc  1400176 Oct 25 19:51 logs/2021122106/gdaspost_f006.log
-rw-r--r-- 1 kate.friedman emc  1400400 Oct 25 19:51 logs/2021122106/gdaspost_f005.log
-rw-r--r-- 1 kate.friedman emc  1400077 Oct 25 19:51 logs/2021122106/gdaspost_f004.log
-rw-r--r-- 1 kate.friedman emc  1400312 Oct 25 19:51 logs/2021122106/gdaspost_f008.log
-rw-r--r-- 1 kate.friedman emc    62640 Oct 25 19:55 logs/2021122106/gdasfit2obs.log
-rw-r--r-- 1 kate.friedman emc  2339596 Oct 25 19:55 logs/2021122106/gdasvrfy.log
-rw-r--r-- 1 kate.friedman emc   162839 Oct 25 20:02 logs/2021122106/gdasarch.log
-rw-r--r-- 1 kate.friedman emc    74445 Oct 25 20:05 logs/2021122106/gdascleanup.log
```

I can still add them to the archive dependency though if we wanna be sure.

Comment (WalterKolczynski-NOAA) [reply workflow/rocoto/gfs_tasks.py:1]: Never depend on a job being done unless it is explicit, no matter how short it is and "should" be done. What happens if it fails?

Comment (KateFriedman-NOAA) [reply workflow/rocoto/gfs_tasks.py:1]: Very true. I have added the GSI monitor jobs to the archival job dependency @ [b7b2638](https://github.com/NOAA-EMC/global-workflow/pull/1983/commits/b7b2638a492986103f6e40e48eeda0fdd0e57f89). Thanks @WalterKolczynski-NOAA !

**Review by WalterKolczynski-NOAA — APPROVED** [review]
Conditionally approved pending successful CI completion.

**Comment by KateFriedman-NOAA** [comment]
The CI test job failure appears unrelated to this PR. The logs shows an unbound `rCDUMP` variable:
```
746 /scratch1/NCEPDEV/global/Terry.McGuinness/GFS_CI_ROOT/PR/1983/global-workflow/scripts/exglobal_stage_ic.sh: line 55: rCDUMP: unbound variable
```
I see the variable is set earlier but not exported because it's "gefs":
```
568 + JGLOBAL_STAGE_IC[8]: rCDUMP=gefs
569 + JGLOBAL_STAGE_IC[9]: [[ gefs = \g\f\s ]]
```
Did the gefs test work in prior PR CI tests?

**Comment by WalterKolczynski-NOAA** [comment]
It passed for #1954, though I'm not sure how. Looks like `JGLOBAL_STAGE_IC` needs to export `rCDUMP`.

**Comment by aerorahul** [comment]
https://github.com/NOAA-EMC/global-workflow/pull/1954/files#r1378000464

**Comment by WalterKolczynski-NOAA** [comment]
> https://github.com/NOAA-EMC/global-workflow/pull/1954/files#r1378000464

Well, too late now, it's already been merged.

**Comment by WalterKolczynski-NOAA** [comment]
I'm making a PR to fix it

**Comment by WalterKolczynski-NOAA** [comment]
`gdasarch` for the half-cycle is not firing because it is waiting for a `gdasverfozn` job that does not exist:
```
  dependencies
    AND is not satisfied
      gdasverfozn of cycle 202112201800 is not SUCCEEDED
```

**Comment by KateFriedman-NOAA** [comment]
> gdasarch for the half-cycle is not firing because it is waiting for a gdasverfozn job that does not exist:

Yeah, so since the GSI monitor jobs don't run in the first half cycle we need the gdasarch job in the first half cycle to not be dependent on them. In lieu of reworking the first half cycle dependencies/jobs entirely...the xml needs something like this:
```
        <dependency>
                <and>
                        <taskdep task="gdasvrfy"/>
                        <taskdep task="gdasfit2obs"/>
                        <or>
                                <and>
                                        <taskdep task="gdasverfozn"/>
                                        <taskdep task="gdasverfrad"/>
                                        <taskdep task="gdasvminmon"/>
                                </and>
                                <not><cycleexistdep cycle_offset="-06:00:00"/></not>
                        </or>
                </and>
        </dependency>
```

Contemplating how to achieve this...

**Comment by WalterKolczynski-NOAA** [comment]
That XML looks correct.

**Comment by KateFriedman-NOAA** [comment]
Have modified the dependencies for the gdas/gfs archive job to account for the lack of GSI monitor jobs in the first half cycle. The xml will now have the dependency logic shown in a [comment above](https://github.com/NOAA-EMC/global-workflow/pull/1983#issuecomment-1791171666) for the gdasarch job. Committed to PR branch @ https://github.com/NOAA-EMC/global-workflow/pull/1983/commits/3ad60053671e3b2bdde3adb3453a77a8fe3d17b2.

When all three GSI monitor jobs are on:
gdasarch:
```
        <dependency>
                <and>
                        <taskdep task="gdasvrfy"/>
                        <taskdep task="gdasfit2obs"/>
                        <or>
                                <and>
                                        <taskdep task="gdasverfozn"/>
                                        <taskdep task="gdasverfrad"/>
                                        <taskdep task="gdasvminmon"/>
                                </and>
                                <not><cycleexistdep cycle_offset="-06:00:00"/></not>
                        </or>
                </and>
        </dependency>
```
gfsarch:
```
        <dependency>
                <and>
                        <taskdep task="gfsvminmon"/>
                        <taskdep task="gfsvrfy"/>
                </and>
        </dependency>
```
When all three GSI monitor jobs are off:
gdasarch:
```
        <dependency>
                <and>
                        <taskdep task="gdasvrfy"/>
                        <taskdep task="gdasfit2obs"/>
                </and>
        </dependency>
```
gfsarch:
```
        <dependency>
                <and>
                        <taskdep task="gfsvrfy"/>
                </and>
        </dependency>
```
When just one GSI monitor job (gdasverfozn) is on:
gdasarch:
```
        <dependency>
                <and>
                        <taskdep task="gdasvrfy"/>
                        <taskdep task="gdasfit2obs"/>
                        <or>
                                <and>
                                        <taskdep task="gdasverfozn"/>
                                </and>
                                <not><cycleexistdep cycle_offset="-06:00:00"/></not>
                        </or>
                </and>
        </dependency>
```

**Comment by WalterKolczynski-NOAA** [comment]
I've asked @TerrenceMcGuinness-NOAA to regenerate the XML so the experiments can proceed without restarting the whole thing.

**Comment by TerrenceMcGuinness-NOAA** [comment]
```
(gfs_workflow) Terry.McGuinness (hfe12) global-workflow (feature/monitor_jobs_separate) $ git pull
remote: Enumerating objects: 29, done.
remote: Counting objects: 100% (29/29), done.
remote: Compressing objects: 100% (19/19), done.
remote: Total 29 (delta 11), reused 24 (delta 10), pack-reused 0
Unpacking objects: 100% (29/29), done.
From https://github.com/KateFriedman-NOAA/global-workflow
 * branch              feature/monitor_jobs_separate -> FETCH_HEAD
Updating 202a2606..3ad60053
Fast-forward
 jobs/JGLOBAL_PREP_OCEAN_OBS         |  10 +++++++++-
 parm/config/gfs/config.prepoceanobs |   8 ++++++--
 scripts/exglobal_archive.sh         |  12 ++++--------
 sorc/link_workflow.sh               |   1 +
 ush/hpssarch_gen.sh                 | 214 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++------------------------------------------------------------------------------------------------------------
 workflow/rocoto/gfs_tasks.py        |  33 +++++++++++++++++++++++----------
 6 files changed, 149 insertions(+), 129 deletions(-)
(gfs_workflow) Terry.McGuinness (hfe12) global-workflow (feature/monitor_jobs_separate) $ 

```

**Comment by KateFriedman-NOAA** [comment]
Thanks @WalterKolczynski-NOAA @TerrenceMcGuinness-NOAA !

**Comment by TerrenceMcGuinness-NOAA** [comment]
```
(gfs_workflow) Terry.McGuinness (hfe12) workflow (feature/monitor_jobs_separate) $ rocotostat -w /scratch1/NCEPDEV/global/Terry.McGuinness/GFS_CI_ROOT/PR/1983/RUNTESTS/EXPDIR/C96C48_hybatmDA_202a2606/C96C48_hybatmDA_202a2606.xml -d /scratch1/NCEPDEV/global/Terry.McGuinness/GFS_CI_ROOT/PR/1983/RUNTESTS/EXPDIR/C96C48_hybatmDA_202a2606/C96C48_hybatmDA_202a2606.db -c 202112201800 -t gdasarch
       CYCLE                    TASK                       JOBID               STATE         EXIT STATUS     TRIES      DURATION
================================================================================================================================
202112201800                gdasarch                    51373636             RUNNING                   -         0           0.0
(gfs_workflow) Terry.McGuinness (hfe12) workflow (feature/monitor_jobs_separate) $ 
```

**Comment by KateFriedman-NOAA** [comment]
I believe this PR is good to go now with the conclusion of the CI testing. Any further comments before merging?

**Pull request merged**
